### PR TITLE
Refactor for improvements and shellcheck

### LINKED ISSUE #17: Fix shellcheck errors and warnings
```
shellcheck genfstab
 
Line 51:
out() { printf "$1 $2\n" "${@:3}"; }
               ^-- SC2059: Don't use variables in the printf format string. Use printf '..%s..' "$foo".
 
Line 72:
  local i= chr= out=
         ^-- SC1007: Remove space after = if trying to assign a value (for empty string, use var='' ... ).
              ^-- SC1007: Remove space after = if trying to assign a value (for empty string, use var='' ... ).
 
Line 90:
  local i= chr= out= len=$(( ${#1} - 4 ))
         ^-- SC1007: Remove space after = if trying to assign a value (for empty string, use var='' ... ).
              ^-- SC1007: Remove space after = if trying to assign a value (for empty string, use var='' ... ).
                   ^-- SC1007: Remove space after = if trying to assign a value (for empty string, use var='' ... ).
 
Line 164:
  local options_=$1
        ^-- SC2178: Variable was used as an array but is now assigned a string.
 
Line 194:
  read dm_name <"/sys/class/block/${1#/dev/}/dm/name"
  ^-- SC2162: read without -r will mangle backslashes.
 
Line 215:
  local src=$1 spec= label= uuid= comment=()
                   ^-- SC1007: Remove space after = if trying to assign a value (for empty string, use var='' ... ).
                          ^-- SC1007: Remove space after = if trying to assign a value (for empty string, use var='' ... ).
                                ^-- SC1007: Remove space after = if trying to assign a value (for empty string, use var='' ... ).
 
Line 243:
  [[ $comment ]] && printf '# %s\n' "${comment[*]}"
     ^-- SC2128: Expanding an array without an index only gives the first element.
 
Line 276:
      if optstring_get_option "$varname" codepage && [[ $codepage = cp* ]]; then
                                                        ^-- SC2154: codepage is referenced but not assigned.
 
Line 417:
  read
  ^-- SC2162: read without -r will mangle backslashes.
```

Reference(s):
- https://github.com/glacion/genfstab/issues/1

### PR DESCRIPTION
**Pull request by YHNdnzj**
Closes #17 

Finally :D

### CODE REVIEW

**Comment by Foxboron** [comment]
I started re-reading the changes here commit by commit, and a tip for the future is to not do refactoring and style changes in the same commit.

https://github.com/archlinux/arch-install-scripts/pull/40/commits/3d86b7ec523bae990a2de2fc9a83dfe7e5de01a3

Is already fairly large, but if you had not done re-ordering and style changes it would have been easier to get an overview of the refactoring work being done.

I do understand parts of this is shellcheck complaining, but that should preferably be separate commits.

File: arch-chroot.in [hunk 1]
```diff
@@ -30,53 +30,6 @@ itself to make it one, i.e. 'mount --bind chroot_dir chroot_dir'.
...
-    # There are two reasons the destination might not exist:
-    #
-    #   1. There may be no resolv.conf in the chroot.  In this case, $dest won't exist,
-    #      and it will be equal to $1/etc/resolv.conf.  In this case, we'll just exit.
-    #      The chroot environment must not be concerned with DNS resolution.
-    #
-    #   2. $1/etc/resolv.conf is (or resolves to) a broken link.  The environment
-    #      clearly intends to handle DNS resolution, but something's wrong.  Maybe it
-    #      normally creates the target at boot time.  We'll (try to) take care of it by
-    #      creating a dummy file at the target, so that we have something to bind to.
```
Comment (Foxboron) [inline arch-chroot.in:65]: These comments are useful, lets not remove them.

Comment (YHNdnzj) [reply arch-chroot.in:65]: Hmm, but I changed the logic to always create `resolv.conf` in chroot. An empty file doesn't hurt I guess?

And if it's a symlink, after my change the symlink would be moved away and we bind mount to a newly-created regular file. The symlink will be restored during cleanup.

Shall I just add a description of the new logic?

**Comment by evelikov** [comment]
Greetings, recently I had a go at fixing some pet peeves in this project. I'm glad to see that you've addressed some of those, although it seems like this PR is far too long and doing multiple unrelated things at the same time.

Will drop some individual notes in a moment, hope you find them useful

File: arch-chroot.in [hunk 2]
```diff
@@ -2,27 +2,30 @@
...
 shopt -s extglob
 
+# m4_include() is recognized as a function definition
+# shellcheck source=common disable=SC1073,SC1065,SC1064,SC1072
 m4_include(common)
 
 setup=chroot_setup
 unshare=0
+userspec=''
+chroot_args=()
```
Comment (evelikov) [inline arch-chroot.in:12]: The commit mixes functional and cosmetic changes. It also introduces promotes local variables to global scope eg `chroot_args` - that should either be voided to documented.

I would suggest fixing the functional/shellcheck issues - one commit per script. Then doing the same on the cosmetic side.

File: arch-chroot.in [hunk 3]
```diff
@@ -30,53 +30,6 @@ itself to make it one, i.e. 'mount --bind chroot_dir chroot_dir'.
...
-    [[ -n $root && $target != $root* ]] && target=$root${target#/}
-  done
-
-  printf %s "$target"
-}
-
-chroot_add_resolv_conf() {
-  local chrootdir=$1
-  local src=$(resolve_link /etc/resolv.conf)
-  local dest=$(resolve_link "$chrootdir/etc/resolv.conf" "$chrootdir")
```
Comment (evelikov) [inline arch-chroot.in:50]: The commit moves some code around while also changes the functionality. These two should be separate commits.

The commit message mentions "chroot_add_resolve_link is cleaned up to be simpler" where chroot_add_resolve_link does not exist in either new or old code. Ideally the message will mention what or why the new approach is simpler.

**Comment by YHNdnzj** [comment]
To-do for myself: add a comment block to the end of the generated fstab, for quirks like https://github.com/archlinux/arch-install-scripts/pull/44

**Comment by klausenbusk** [comment]
This has been migrated to: https://gitlab.archlinux.org/archlinux/arch-install-scripts/-/merge_requests/40.
